src/runtime/resolver.rs

```diff
@@ -0,0 +1,159 @@
+use std::{future::Future, net::SocketAddr, time::Duration};
+
+use async_trait::async_trait;
+use futures_timer::Delay;
+use trust_dns_proto::error::ProtoError;
+use trust_dns_resolver::{
+    lookup::{SrvLookup, TxtLookup},
+    name_server::{GenericConnection, GenericConnectionProvider},
+    IntoName,
+};
+
+use crate::{
+    error::{Error, Result},
+    runtime::{stream::AsyncTcpStream, AsyncRuntime},
+    RUNTIME,
+};
+
+// Some shorter type names for the generic constraints of the underlying trust-dns
+// `AsyncResolver`.
+type AsyncConnectionProvider = GenericConnectionProvider<AsyncRuntimeProvider>;
+type TrustDnsResolver =
+    trust_dns_resolver::AsyncResolver<GenericConnection, AsyncConnectionProvider>;
+
+/// An async runtime agnostic DNS resolver.
+pub(crate) struct AsyncResolver {
+    resolver: TrustDnsResolver,
+}
+
+impl AsyncResolver {
+    pub(crate) async fn new() -> Result<Self> {
+        let resolver = TrustDnsResolver::from_system_conf(crate::RUNTIME).await?;
+        Ok(Self { resolver })
+    }
+}
+
+impl AsyncResolver {
+    pub async fn srv_lookup<N: IntoName>(&self, query: N) -> Result<SrvLookup> {
+        let lookup = self.resolver.srv_lookup(query).await?;
+        Ok(lookup)
+    }
+
+    pub async fn txt_lookup<N: IntoName>(&self, query: N) -> Result<TxtLookup> {
+        let lookup = self.resolver.txt_lookup(query).await?;
+        Ok(lookup)
+    }
+}
+
+/// The provider type used by trustdns to operate using the proper async runtime.
+#[derive(Copy, Clone)]
+struct AsyncRuntimeProvider {}
```

Comment:
Interestingly, in Rust there's a slight difference between a struct with zero fields and a singleton struct; in this case, I think it's more idiomatic to use the latter (i.e. `struct AsyncRuntimeProvider;` rather than `struct AsyncRuntimeProvider {}`).